### PR TITLE
USEID-737: Add time based tokens

### PR DESCRIPTION
**Pull request by code28**

### CODE REVIEW

File: src/main/kotlin/de/bund/digitalservice/useid/config/SecurityConfig.kt [hunk 1]
```diff
@@ -22,6 +22,8 @@ class SecurityConfig(
     fun springSecurityFilterChain(http: ServerHttpSecurity): SecurityWebFilterChain {
         return http.authorizeExchange()
             .pathMatchers("$IDENTIFICATION_SESSIONS_BASE_PATH/*/tc-token").permitAll()
+            .pathMatchers("$IDENTIFICATION_SESSIONS_BASE_PATH/*/tokens/*").permitAll()
+            .pathMatchers("$IDENTIFICATION_SESSIONS_BASE_PATH/*/tokens").permitAll()
```
Comment (code28) [inline src/main/kotlin/de/bund/digitalservice/useid/config/SecurityConfig.kt:26]: @mpanne If you have an idea to merge those two, I'd be happy for help... (`tokens*` or `tokens**` didn't work)

File: src/main/kotlin/de/bund/digitalservice/useid/timebasedtokens/TimeBasedTokenController.kt [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+import reactor.core.publisher.Mono
+import java.util.UUID
+
+@RestController
+@Timed
+@RequestMapping(IDENTIFICATION_SESSIONS_BASE_PATH)
+@Tag(name = "Time-based Tokens", description = "Time-based tokens inside the widget that are only 60 seconds valid and can be checked by the app.")
+@ConditionalOnProperty(name = ["features.desktop-solution-enabled"], havingValue = "true")
+class TimeBasedTokenController(private val timeBasedTokenService: TimeBasedTokenService) {
+    private val log = KotlinLogging.logger {}
```
Comment (mpanne) [inline src/main/kotlin/de/bund/digitalservice/useid/timebasedtokens/TimeBasedTokenController.kt:27]: logger is never used.

File: src/main/kotlin/de/bund/digitalservice/useid/timebasedtokens/TimeBasedTokenService.kt [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+                        log.error("Error when deleting time based token", it)
+                        Mono.empty<Void>()
+                    }
+                    .then(mockDatasource.save(TimeBasedToken(useIdSessionId, UUID.randomUUID())))
+            }
+    }
+
+    fun isTokenValid(useIdSessionId: UUID, tokenId: UUID): Mono<Boolean> {
+        return mockDatasource.findByUseIdSessionIdAndTokenId(useIdSessionId, tokenId)
+            .map { it?.let { it.createdAt != null && it.createdAt!!.plusSeconds(60) > LocalDateTime.now() } }
```
Comment (mpanne) [inline src/main/kotlin/de/bund/digitalservice/useid/timebasedtokens/TimeBasedTokenService.kt:33]: IntelliJ suggests to use `mapNotNull` here.

**Review by mpanne — APPROVED** [review]
Looks good!

File: src/main/kotlin/de/bund/digitalservice/useid/config/SecurityConfig.kt (hunk 1 above)
Comment (code28) [reply src/main/kotlin/de/bund/digitalservice/useid/config/SecurityConfig.kt:26]: Good point 👍

File: src/main/kotlin/de/bund/digitalservice/useid/timebasedtokens/TBTMockDatasource.kt [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+        timeBasedTokens.add(timeBasedToken)
+        return Mono.just(timeBasedToken)
+    }
+
+    fun deleteAllByUseIdSessionId(useIdSessionId: UUID): Mono<Void> {
+        timeBasedTokens.removeAll { it.useIdSessionId == useIdSessionId }
+        return Mono.empty()
+    }
+
+    fun findByUseIdSessionIdAndTokenId(useIdSessionId: UUID, tokenId: UUID): Mono<TimeBasedToken?> {
```
Comment (mpanne) [inline src/main/kotlin/de/bund/digitalservice/useid/timebasedtokens/TBTMockDatasource.kt:25]: Why is there a `?` in the return type? The time based token will never be null right? Either a tbt is found and returned wrapped by `Mono`, or an empty `Mono` is returned. Would make the code of the caller easier.

**Pull request merged**